qiskit_nature/optionals.py

```diff
@@ -0,0 +1,90 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Additional optional constants.
+"""
+
+from typing import Optional, Callable
+from shutil import which
+from qiskit.utils import LazyDependencyManager, LazyImportTester, LazySubprocessTester
+
+
+class NatureLazyCommandTester(LazyDependencyManager):
+    """
+    A lazy checker that a command-line tool is available. This just checks if the command is
+    not empty.
+    """
+
+    __slots__ = ("_command",)
+
+    def __init__(
+        self,
+        command: str,
+        *,
+        name: Optional[str] = None,
+        callback: Optional[Callable[[bool], None]] = None,
+        install: Optional[str] = None,
+        msg: Optional[str] = None,
+    ):
+        """
+        Args:
+            command: the string that make up the command.  For example,``"g16"``.
+        """
+        self._command = "" if command is None else command
+        super().__init__(name=name or self._command, callback=callback, install=install, msg=msg)
+
+    def _is_available(self):
+        return bool(self._command)
+
+
+HAS_PYQUANTE2 = LazyImportTester(
+    {
+        "pyquante2": ("molecule", "rhf", "uhf", "rohf", "basisset", "onee_integrals"),
+        "pyquante2.geo.zmatrix": ("z2xyz",),
+        "pyquante2.ints.integrals": ("twoe_integrals",),
+    },
+    name="pyquante2",
+    msg="See https://github.com/rpmuller/pyquante2",
+)
+
+HAS_PYSCF = LazyImportTester(
+    {
+        "pyscf": ("__version__", "gto", "scf"),
+        "pyscf.lib": ("chkfile", "logger", "param"),
+        "pyscf.tools": ("dump_mat",),
+    },
+    name="pyscf",
+    msg="See https://pyscf.org/install.html",
+)
+
+GAUSSIAN_16 = "g16"
+GAUSSIAN_16_DESC = "Gaussian 16"
+G16PROG = which(GAUSSIAN_16)
+
+HAS_GAUSSIAN = NatureLazyCommandTester(
+    G16PROG,
+    name=GAUSSIAN_16_DESC,
+    msg="Please check that it is installed correctly",
+)
+
+PSI4 = "psi4"
+PSI4_DESC = "PSI4"
+PSI4PROG = which(PSI4)
+if PSI4PROG is None:
+    PSI4PROG = PSI4
```

Comment:
Why is this needed? If `which` cannot find the command how would we be able to use it?

Reply:
@mrossinek Just because the LazySubprocessTester from Terra won't accept a None or empty command in the constructor and I still wanted to use its capabilities of actually running the command to see if it is installed.

Reply:
Okay, then why is the same not being done in the gaussian case a few lines further up?

Reply:
Ahhh, because unfortunately Gaussian doesn't have a command line option like version or whatever that lets me just simply run it without side effect of actually running code. For this I created my own Lazy class derived from Terra that just checks if the command is None. Effectively just what was being done before: 
```
class NatureLazyCommandTester(LazyDependencyManager):
```

Reply:
Oh I was not careful. The first time I read it, it looked like you used `NatureLazyCommandTester` for Gaussian and Psi4. But I see now that the latter actually uses the Terra-provided one :+1:

Reply:
What functionality would you like in `LazySubprocessTester`?  I made it reject "empty" commands as a way of catching bugs because I couldn't think of any use-cases, but if you've got a use-case for empty commands, we could add it.

Here, I would note that calling `which` on import isn't lazy - if you wanted it for your Gaussian one, perhaps your derived class could run `shutil.which` to define its `_is_available`?  It depends how lazy you care about being, though.